File: system-test/dotnet/Common/Result.cs
namespace Common;

public class Result<T, E>
{
    private readonly bool _success;
    private readonly T? _value;
    private readonly E? _error;

    private Result(bool success, T? value, E? error)
    {
        _success = success;
        _value = value;
        _error = error;
    }

    public static Result<T, E> Success(T value) => new(true, value, default);

    public static Result<T, E> Failure(E error) => new(false, default, error);

    public bool IsSuccess => _success;

    public bool IsFailure => !_success;

    public T Value
    {
        get
        {
            if (!_success)
                throw new InvalidOperationException("Cannot get value from a failed result");
            return _value!;
        }
    }

    public E Error
    {
        get
        {
            if (_success)
                throw new InvalidOperationException("Cannot get error from a successful result");
            return _error!;
        }
    }

    public Result<T2, E> Map<T2>(Func<T, T2> mapper)
    {
        if (_success)
        {
            return Result<T2, E>.Success(mapper(_value!));
        }
        return Result<T2, E>.Failure(_error!);
    }

    public Result<T, E2> MapError<E2>(Func<E, E2> mapper)
    {
        if (_success)
        {
            return Result<T, E2>.Success(_value!);
        }
        return Result<T, E2>.Failure(mapper(_error!));
    }

    public Result<VoidValue, E> MapVoid()
    {
        if (_success)
        {
            return Result<VoidValue, E>.Success(VoidValue.Empty);
        }
        return Result<VoidValue, E>.Failure(_error!);
    }
}

public static class Result
{
    public static Result<VoidValue, E> Success<E>()
        => Result<VoidValue, E>.Success(VoidValue.Empty);

    public static Result<VoidValue, E> Failure<E>(E error)
        => Result<VoidValue, E>.Failure(error);
}

File: system-test/java/src/main/java/com/mycompany/myshop/testkit/common/Result.java
package com.mycompany.myshop.testkit.common;

public class Result<T, E> {
    private final boolean success;
    private final T value;
    private final E error;

    private Result(boolean success, T value, E error) {
        this.success = success;
        this.value = value;
        this.error = error;
    }

    public static <T, E> Result<T, E> success(T value) {
        return new Result<>(true, value, null);
    }

    public static <T, E> Result<T, E> failure(E error) {
        return new Result<>(false, null, error);
    }

    public static <E> Result<Void, E> success() {
        return new Result<>(true, null, null);
    }

    public boolean isSuccess() {
        return success;
    }

    public boolean isFailure() {
        return !success;
    }

    public T getValue() {
        if (!success) {
            throw new IllegalStateException("Cannot get value from a failed result");
        }
        return value;
    }

    public E getError() {
        if (success) {
            throw new IllegalStateException("Cannot get error from a successful result");
        }
        return error;
    }

    public <T2> Result<T2, E> map(java.util.function.Function<T, T2> mapper) {
        if (success) {
            return Result.success(mapper.apply(value));
        }
        return Result.failure(error);
    }

    public <E2> Result<T, E2> mapError(java.util.function.Function<E, E2> mapper) {
        if (success) {
            return Result.success(value);
        }
        return Result.failure(mapper.apply(error));
    }

    public Result<Void, E> mapVoid() {
        if (success) {
            return Result.success();
        }
        return Result.failure(error);
    }
}


